### PR TITLE
[ONNX] Simplify repeat_intereleave export for scalar-valued 'repeat'

### LINKED ISSUE #100438: [ONNX] torch.repeat_interleave() export creates slow ONNX models
### 🐛 Describe the bug

I have noticed that `torch.repeat_interleave` creates ONNX models which have much more ops than they need to. Please consider the following code, where I export a model with a single repeat_interleave op to ONNX:

```python
import torch
import torch.nn as nn

args = (torch.rand((4, 4, 32)),)

# Model which uses repeat_interleave
class RepeatInterleaveModel(nn.Module):
    def forward(self, x):
        return x.repeat_interleave(2, dim=-1)

model = RepeatInterleaveModel()

# Resulting onnx model has 463 nodes!
torch.onnx.export(model, args, "repeat_interleave.onnx", opset_version=17)
```

The resulting ONNX model has **463** nodes, which is way too many. 

Fewer ops are produced if we make the last axis dynamic:
```python
torch.onnx.export(model, args, "dynamic_repeat_interleave.onnx", opset_version=17,
                  input_names=['input'], output_names=['output'], dynamic_axes={'input': [2]})
```
The created ONNX model has a `Loop` op which creates the output tensor by using `Sequence` ops. This is correct but pretty inefficient - because we are repeating each element the same number of times (as is often the case), we can create the output simply by doing a `Unsqueeze` + `Expand` on the input. It would be best if possible to avoid creating the `Loop`, since it is much slower run during inference (compared to the `Expand`), and will also make it hard to do shape inference of the model's tensors. Below I have a pytorch model which does the same thing as the original model but has a much cleaner ONNX export:
```python
# Model which produces the same result but with fewer nodes.
class UnsqueezeRepeat(nn.Module):
    def forward(self, x):
        repeats = tuple((1,) * len(x.shape) + (2,))
        return x.unsqueeze(-1).repeat(repeats).flatten(-2, -1)

alt_model = UnsqueezeRepeat()
assert torch.equal(model(*args), alt_model(*args)) # Check correctness

# Exports to only 15 nodes. Exported graph is the same for both static and dynamic versions.
torch.onnx.export(alt_model, args, "alt_repeat_interleave.onnx", opset_version=17)
torch.onnx.export(alt_model, args, "alt_dynamic_repeat_interleave.onnx", opset_version=17,
                  input_names=['input'], output_names=['output'], dynamic_axes={'input': [2]})
```
Exported ONNX:
clean_onnx

In Summary, when the `repeats` parameter to `torch.repeat_interleave` is an int (meaning that each element is to be repeated the same number of times), this can be implemented in ONNX using an `Expand` op. Skimming the code in torch.onnx, I think this is the intention, but maybe some bugs are preventing it from working 100%.

Thanks!

### Versions

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.11.1
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] mypy==0.960
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.1
[pip3] torch==2.1.0a0+git0cf6e74
[pip3] torchvision==0.16.0.dev20230330+cpu
[conda] magma-cuda118             2.6.1                         1    pytorch
[conda] mkl                       2022.1.0           hc2b9512_224  
[conda] mkl-include               2022.1.0           h06a4308_224  
[conda] numpy                     1.23.1                   pypi_0    pypi
[conda] torch                     2.1.0a0+git0cf6e74           dev_0    <develop>
[conda] torchvision               0.16.0.dev20230330+cpu          pypi_0    pypi

### PR DESCRIPTION
**Pull request by ilyasher**
This PR simplifies the ONNX export of torch.repeat_interleave when 'repeat' is a scalar value (so each index in the input is repeated the same number of times). (Issue #100438)

Here is a before/after of a simple model export:
```python
# Model + export code
import torch

class RepeatInterleaveModel(torch.nn.Module):
    def forward(self, x):
        return x.repeat_interleave(2, dim=-1)

args = (torch.rand((2, 2, 16)),)
model = RepeatInterleaveModel()
torch.onnx.export(model, args, "repeat_interleave.onnx", opset_version=17)
``` 

**Before (static shapes)**
repeat_interleave onnx(1)

-----
**Before (dynamic shapes, second graph is Loop body)**
<p float="left">
  
   
</p>

-----
**After (for both static and dynamic shapes)**

-----

This PR also fixes a bug where the exporter throws an expection when the input has dynamic shapes and the 'dim' parameter is not specified to torch.repeat_interleave. Also adds a new testcase to cover this. (Issue #100429)

Fixes #100438 and #100429

### CODE REVIEW

**Comment by BowenBao** [comment]
Could possibly fix #98077 too.

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -1262,6 +1262,49 @@ def _repeat_interleave_split_helper(g: jit_utils.GraphContext, self, reps, dim):
     return split_out if reps > 1 else [split_out]
 
 
+@_beartype.beartype
+def _repeat_interleave_single_value_helper(
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:1266]: nit:
```suggestion
def _repeat_interleave_single_value_repeat_helper(
```
or
```suggestion
def _repeat_interleave_single_repeat_helper(
```

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -1262,6 +1262,49 @@ def _repeat_interleave_split_helper(g: jit_utils.GraphContext, self, reps, dim):
...
+        repeats = g.op("Constant", value_t=torch.LongTensor(repeats))
+
+    const_repeats: bool = _is_constant(repeats)
+    reps = _maybe_get_const(repeats, "t")
+
+    # Convert 'repeats' to 1-d if it is 0-d.
+    if _get_tensor_rank(repeats) == 0:
+        repeats = g.op("Reshape", repeats, g.op("Constant", value_t=torch.tensor([1])))
+
+    # Create a new dim of size 1, then resize it to be 'repeats' long, and finally collapse it.
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py:1281]: nit: expand or tile might be more precise.
```suggestion
    # Create a new dim of size 1, then expand it to be 'repeats' long, and finally collapse it.
```

**Review by BowenBao — COMMENTED** [review]
Looks good overall w/ minor comments. Thanks for contributing! 🚀

**Review by BowenBao — APPROVED** [review]

**Comment by ilyasher** [comment]
@BowenBao I am ready to merge, anything else left to do? Thanks!

**Comment by BowenBao** [comment]
@ilyasher merged. Thanks again!
